text/inductive-flexible-parameters-indices.md

```diff
@@ -0,0 +1,128 @@
+- Title: Flexible combination of parameters and indices in inductive types
+
+- Drivers: Hugo
+
+----
+
+# Summary
+
+Inductive types have three kinds of arguments:
+- uniform parameters
+- recursively non-uniform parameters
+- indices (also known as real arguments)
+
+These arguments have to be given by blocks in this order. The CEP proposes to unify the syntax of parameters and indices and to let Coq compute automatically the status of an argument, accepting combining arguments of different status in any order.
+
+This has the following advantages:
+- the `:` in the type of inductive type behaves like the `:` in declarations, that is, `(x : A) :` is the same as `: forall x :
+A,`
+- no more constraint on giving parameters before the indices and the uniform parameters before the recursively non-uniform ones
+- different implicit status can be given at definition time to parameters depending on whether it is the parameter of a constructor or of the inductive type
+- no more difference of status between an argument given as a parameter and the same kind of argument given as a index
+- not all inductive types of a block of mutual inductive types need to have exactly the same parameters
+
+# Basic proposal
+
+The basic proposed changes are:
+- the context of parameters currently given next to the inductive type can optionally be replicated in the type of constructors
+- if the context is replicated in all constructors, it can be dropped for the inductive type itself
+- any argument declared after the `:` in the type of the inductive type can equivalently be moved before the `:` and the parameter/index status is instead computed automatically
+
+# Examples
+
+The following syntaxes all define the same `list A` where `A` is a parameter:
+```
+Inductive list (A : Type) : Type :=
+| nil : list A
+| cons : A -> list A -> list A.
+
+Inductive list : forall (A : Type), Type :=
+| nil : list A
+| cons : A -> list A -> list A.
+
+Inductive list : forall (A : Type), Type :=
+| nil : forall A, list A
+| cons : forall A, A -> list A -> list A.
```

Comment:
This one and the `Inductive list := ...` one at the end make sense IMO, but the ones were only some of the constructors get the `forall` and the ones where `A` is both on the left of the `:` and as a `forall` in a constructor don't.
This is similar to definitions: if I do `Definition foo (x:bla) : nat := fun x => 0` I get an error.

Reply:
Yes, this is the remark "For clarity of scoping, one may also continue to impose that a variable occurring after the `:` cannot be used in the type of constructors."